File: packages/sdk/src/namespaces/physics.ts
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */

/**
 * bim.physics — rigid-body what-if simulations.
 *
 * Usage:
 *   const result = bim.physics.simulate({ remove: [columnRef.expressId] });
 *   bim.viewer.colorize(result.falling.map(id => ({ modelId, expressId: id })), '#dc2626');
 *
 * Plausibility check, not structural engineering — no bending or material
 * yield. See `@ifc-lite/physics` for the full contract.
 */

import type {
  BimBackend,
  PhysicsSimulateOptions,
  PhysicsSimulationResult,
} from '../types.js';

export class PhysicsNamespace {
  constructor(private backend: BimBackend) {}

  /**
   * Run a rigid-body simulation against the active model (or the named one).
   *
   * Single-model only at the moment — federated multi-model simulation is
   * intentionally out of scope for v1.
   */
  simulate(options?: PhysicsSimulateOptions): PhysicsSimulationResult;
  simulate(modelId: string, options?: PhysicsSimulateOptions): PhysicsSimulationResult;
  simulate(
    arg1?: string | PhysicsSimulateOptions,
    arg2?: PhysicsSimulateOptions,
  ): PhysicsSimulationResult {
    if (typeof arg1 === 'string') {
      return this.backend.physics.simulate(arg1, arg2 ?? {});
    }
    return this.backend.physics.simulate(null, arg1 ?? {});
  }
}

File: rust/engine/src/physics.rs
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at https://mozilla.org/MPL/2.0/.

//! Engine-facade entry point for rigid-body what-if simulations.
//!
//! Wraps `ifc-lite-physics` so callers (server, desktop, CLI) can pass the
//! `MeshData` they already have without converting to a third type.
//!
//! See `ifc-lite-physics` for the simulation contract — this module only
//! adapts inputs.

use ifc_lite_physics::{simulate as physics_simulate, PhysicsMesh};
use ifc_lite_processing::MeshData;

pub use ifc_lite_physics::{
    Aabb as PhysicsAabb, AnchorReason, BodyOutcome, SimulateOptions, SimulationResult, Stability,
};

/// Run a rigid-body simulation against the engine's `MeshData` shape.
///
/// `meshes` is consumed read-only; only `express_id`, `ifc_type`, `positions`,
/// and `indices` are read.
pub fn simulate(meshes: &[MeshData], options: &SimulateOptions) -> SimulationResult {
    let physics_meshes: Vec<PhysicsMesh> = meshes
        .iter()
        .filter(|m| !m.is_empty())
        .map(|m| PhysicsMesh {
            express_id: m.express_id,
            ifc_type: m.ifc_type.clone(),
            positions: m.positions.clone(),
            indices: m.indices.clone(),
        })
        .collect();
    physics_simulate(&physics_meshes, options)
}
